beam/tools/src/test/java/cz/o2/proxima/beam/tools/groovy/BeamStreamTest.java

```diff
@@ -610,6 +630,66 @@ public void testExceptionRethrow() {
     expectThrow(() -> BeamStream.rethrow(new IOException()), IllegalStateException.class);
   }
 
+  @Test
+  public void testImpulse() {
+    WindowedStream<Integer> impulse = BeamStream.impulse(op, Pipeline::create, () -> 1);
+    List<Pair<String, Integer>> result =
+        impulse
+            .combine(
+                Closures.from(this, ign -> ""), 0, Closures.from(this, (a, b) -> (int) a + (int) b))
+            .collect();
+    assertEquals(Collections.singletonList(Pair.of("", 1)), result);
+  }
+
+  @Test(timeout = 15000)
+  public void testPeriodicImpulse() throws InterruptedException {
+    SerializableScopedValue<Integer, AtomicBoolean> finished =
+        new SerializableScopedValue<>(AtomicBoolean::new);
+    SerializableScopedValue<Integer, AtomicInteger> seen =
+        new SerializableScopedValue<>(AtomicInteger::new);
+    SerializableScopedValue<Integer, AtomicInteger> emitted =
+        new SerializableScopedValue<>(AtomicInteger::new);
+    WindowedStream<Integer> impulse =
+        BeamStream.periodicImpulse(
+            op,
+            () -> {
+              PipelineOptions opts = PipelineOptionsFactory.create();
+              opts.as(DirectOptions.class).setBlockOnRun(false);
+              return Pipeline.create(opts);
+            },
+            () -> emitted.get(0).incrementAndGet(),
+            1000,
+            () -> finished.get(0).get());
+
+    Thread pipeline =
+        new Thread(
+            () -> {
+              try {
+                impulse
+                    .map(
+                        Closures.from(
+                            this,
+                            el -> {
+                              if (seen.get(0).incrementAndGet() > 3) {
+                                finished.get(0).set(true);
+                              }
+                              return el;
+                            }))
+                    .collect();
+              } catch (Exception ex) {
+                if (ex.getCause() != null && ExceptionUtils.isInterrupted(ex.getCause())) {
+                  // pass
+                }
+                throw ex;
+              }
+            });
+    pipeline.setDaemon(true);
+    pipeline.start();
+    while (emitted.get(0).get() < 3) {
```

Comment:
nit: Personally, I prefer `CountDownLatch` instead of busy wait

Reply:
makes sense :+1: